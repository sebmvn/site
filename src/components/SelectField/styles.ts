import styled, { css, DefaultTheme } from 'styled-components'

import { SelectFieldProps } from '.'

type IconPositionProps = Pick<SelectFieldProps, 'iconPosition'>

type WrapperProps = Pick<SelectFieldProps, 'disabled'> & { error?: boolean }

export const Option = styled.option`
  ${({ theme }) => css`
    background: ${theme.colors.inputBg};
  `}
`

export const SelectWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    background: ${theme.colors.secondary};
    border-radius: ${theme.border.radius};
    padding: 0 ${theme.spacings.xsmall};
    border-color: ${theme.colors.inputBg};

    cursor: pointer;

    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }
  `}
`

export const Select = styled.select<IconPositionProps>`
  ${({ theme, iconPosition }) => css`
    flex: 1;
    width: ${iconPosition === 'right' ? `calc(100% - 2.2rem)` : ''};
    
    -webkit-appearance: button;
    cursor: pointer;

    color: ${theme.colors.text};
    background: ${theme.colors.secondary};
    border: 0;
    outline: 0;
    box-shadow: none;

    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} 0;
    padding-${iconPosition}: ${theme.spacings.xsmall};



    &:-webkit-autofill {
      -webkit-box-shadow: 0 0 0 ${theme.spacings.small} ${theme.colors.inputBg}
        inset;
      filter: none;
      &::first-line {
        font-family: ${theme.font.family};
        font-size: ${theme.font.sizes.medium};
      }
    }

    &:focus{
      border: 0;
    }

 

  `}
`

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.primary};
    cursor: pointer;
  `}
`

export const Icon = styled.div<IconPositionProps>`
  ${({ theme, iconPosition }) => css`
    display: flex;
    color: ${theme.colors.primary};
    order: ${iconPosition === 'right' ? 1 : 0};

    & > svg {
      width: 2.2rem;
      height: 100%;
    }

    span {
      padding: 0;
    }
  `}
`

export const Error = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.alert};
    font-size: ${theme.font.sizes.xsmall};
  `}
`

const wrapperModifiers = {
  error: (theme: DefaultTheme) => css`
    ${SelectWrapper} {
      border-color: ${theme.colors.alert};
    }

    ${Label} {
      color: ${theme.colors.alert};
    }
  `,
  disabled: (theme: DefaultTheme) => css`
    ${Label},
    ${Select},
    ${Icon} {
      cursor: not-allowed;
      color: ${theme.colors.text}72;

      &::placeholder {
        color: currentColor;
      }
    }
  `
}

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, error, disabled }) => css`
    ${error && wrapperModifiers.error(theme)}
    ${disabled && wrapperModifiers.disabled(theme)}
  `}
`
