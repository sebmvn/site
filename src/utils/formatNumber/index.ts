export const formatPriceNotation = (value: string) => {
  value = value?.replace('.', '').replace(',', '').replace(/\D/g, '')

  if (isNaN(parseFloat(value))) {
    return ''
  }

  const options = { minimumFractionDigits: 2 }
  const result = new Intl.NumberFormat('pt-BR', options).format(
    parseFloat(value) / 100
  )

  return result
}

export const formatIntegerNotation = (value: string) => {
  return value?.replace('.', '').replace(',', '').replace(/\D/g, '')
}

export const toDecimal = (value: string) => {
  const formattedValue = formatIntegerNotation(value)
  return parseFloat(formattedValue!) / 100
}

export const toInteger = (value: string) => {
  const formattedValue = formatIntegerNotation(value)
  return parseInt(formattedValue!)
}
