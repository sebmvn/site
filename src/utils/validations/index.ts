import { CalendarType } from 'hooks/use-calc'
import Joi from 'joi'

const fieldsValidations = {
  initialValue: Joi.string().required().messages({
    'any.required': 'Campo obrigatório',
    'string.empty': 'Campo obrigatório',
    'string.min': 'tamanho minimo de 1 digito',
    'string.base': 'o tipo do dado precisa ser textual'
  }),
  monthlyValue: Joi.string().required().messages({
    'any.required': 'Campo obrigatório',
    'string.empty': 'Campo obrigatório',
    'string.min': 'tamanho minimo de 1 digito',
    'string.base': 'o tipo do dado precisa ser textual'
  }),

  interestRate: Joi.string().required().messages({
    'any.required': 'Campo obrigatório',
    'string.empty': 'Campo obrigatório',
    'string.min': 'tamanho minimo de 1 digito',
    'string.base': 'o tipo do dado precisa ser textual'
  }),
  period: Joi.string().required().messages({
    'any.required': 'Campo obrigatório',
    'string.empty': 'Campo obrigatório',
    'string.min': 'tamanho minimo de 1 digito',
    'string.base': 'o tipo do dado precisa ser textual'
  }),
  interestRateCalendar: Joi.string().valid('monthly', 'yearly').required(),
  periodCalendar: Joi.string().valid('monthly', 'yearly').required()
}

export type FieldErrors = {
  [key: string]: string
}

const getFieldErrors = (objError: Joi.ValidationResult) => {
  const errors: FieldErrors = {}

  if (objError.error) {
    objError.error.details.forEach((err) => {
      errors[err.path.join('.')] = err.message
    })
  }

  return errors
}

type compoundInterestCalculatorFields = {
  initialValue: string
  monthlyValue: string
  interestRate: string
  period: string
  interestRateCalendar: CalendarType
  periodCalendar: CalendarType
}
export const compoundInterestCalculatorValidate = (
  values: compoundInterestCalculatorFields
) => {
  const schema = Joi.object(fieldsValidations)

  return getFieldErrors(schema.validate(values, { abortEarly: false }))
}
