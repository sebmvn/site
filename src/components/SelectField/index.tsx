import { useState, SelectHTMLAttributes } from 'react'

import * as S from './styles'

export type OptionType = {
  value: string
  label: string
}

export type SelectFieldProps = {
  options: OptionType[]
  onInputChange?: (value: string) => void
  label?: string
  initialValue?: string
  icon?: React.ReactNode
  iconPosition?: 'left' | 'right'
  disabled?: boolean
  error?: string
} & SelectHTMLAttributes<HTMLSelectElement>

const SelectField = ({
  options,
  icon,
  iconPosition = 'left',
  label,
  name,
  initialValue = '',
  error,
  disabled = false,
  onInputChange,
  ...props
}: SelectFieldProps) => {
  const [value, setValue] = useState(initialValue)

  const onChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    const newValue = e.currentTarget.value
    setValue(newValue)

    !!onInputChange && onInputChange(newValue)
  }

  return (
    <S.Wrapper disabled={disabled} error={!!error}>
      {!!label && <S.Label htmlFor={name}>{label}</S.Label>}
      <S.SelectWrapper>
        {!!icon && <S.Icon iconPosition={iconPosition}>{icon}</S.Icon>}
        <S.Select
          onChange={onChange}
          iconPosition={iconPosition}
          disabled={disabled}
          name={name}
          value={value}
          {...(label ? { id: name } : {})}
          {...props}
        >
          {options.map(({ value, label }, index) => (
            <S.Option key={index} value={value}>
              {label}
            </S.Option>
          ))}
        </S.Select>
      </S.SelectWrapper>
      {!!error && <S.Error>{error}</S.Error>}
    </S.Wrapper>
  )
}

export default SelectField
