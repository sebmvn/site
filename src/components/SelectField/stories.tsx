import { Story, Meta } from '@storybook/react/types-6-0'
import { Mail } from '@styled-icons/feather/Mail'

import SelectField, { SelectFieldProps } from '.'

const options = [
  {
    value: 'Option 1',
    label: 'Option 1'
  },
  {
    value: 'Option 2',
    label: 'Option 2'
  },
  {
    value: 'Option 3',
    label: 'Option 3'
  }
]

export default {
  title: 'Form/SelectField',
  component: SelectField,
  args: {
    label: 'E-mail',
    name: 'email',
    icon: <Mail />,
    initialValue: '',
    disabled: false,
    options
  },
  argTypes: {
    onInput: { action: 'changed' },
    icon: { type: 'symbol' }
  }
} as Meta<SelectFieldProps>

export const Default: Story<SelectFieldProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <SelectField {...args} />
  </div>
)

export const withError: Story<SelectFieldProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <SelectField {...args} />
  </div>
)

withError.args = {
  error: 'Ops...something is wrong'
}
