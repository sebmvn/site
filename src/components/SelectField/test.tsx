import userEvent from '@testing-library/user-event'
import { Mail } from '@styled-icons/feather/Mail'

import SelectField from '.'
import { renderWithTheme } from 'utils/test-utils'
import { screen } from '@testing-library/react'

describe('<SelectField />', () => {
  const options = [{ value: '1', label: '1' }]

  it('Renders with Label', () => {
    renderWithTheme(
      <SelectField label="Label" name="Label" options={options} />
    )

    expect(screen.getByLabelText('Label')).toBeInTheDocument()
  })

  it('Renders without Label', () => {
    renderWithTheme(<SelectField options={options} />)

    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument()
  })

  it('Renders with placeholder', () => {
    renderWithTheme(<SelectField placeholder="hey you" options={options} />)

    expect(screen.getByPlaceholderText('hey you')).toBeInTheDocument()
  })

  it('Renders with Icon', () => {
    renderWithTheme(
      <SelectField icon={<Mail data-testid="icon" />} options={options} />
    )

    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('Renders with Icon on the right side', () => {
    renderWithTheme(
      <SelectField
        icon={<Mail data-testid="icon" />}
        iconPosition="right"
        options={options}
      />
    )

    expect(screen.getByTestId('icon').parentElement).toHaveStyle({ order: 1 })
  })

  it('Does not changes its value when disabled', async () => {
    const onInputChange = jest.fn()
    renderWithTheme(
      <SelectField
        onInputChange={onInputChange}
        label="TextField"
        name="TextField"
        options={options}
        disabled
      />
    )
  })

  it('Renders with error', () => {
    const { container } = renderWithTheme(
      <SelectField
        icon={<Mail data-testid="icon" />}
        label="TextField"
        error="Error message"
        options={options}
      />
    )

    expect(screen.getByText('Error message')).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('Is accessible by tab', () => {
    renderWithTheme(
      <SelectField label="TextField" name="TextField" options={options} />
    )

    const input = screen.getByLabelText('TextField')
    expect(document.body).toHaveFocus()

    userEvent.tab()
    expect(input).toHaveFocus()
  })

  it('Is not accessible by tab when disabled', () => {
    renderWithTheme(
      <SelectField
        label="TextField"
        name="TextField"
        disabled
        options={options}
      />
    )

    const input = screen.getByLabelText('TextField')
    expect(document.body).toHaveFocus()

    userEvent.tab()
    expect(input).not.toHaveFocus()
  })
})
