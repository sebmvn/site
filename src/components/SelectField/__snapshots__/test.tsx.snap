// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectField /> Renders with error 1`] = `
.c6 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c8 {
  background: #F4F4F4;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #0ADC7D;
  border-radius: 0.5rem;
  padding: 0 1.6rem;
  border-color: #F4F4F4;
  cursor: pointer;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #193241;
}

.c7 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-appearance: button;
  cursor: pointer;
  color: #193241;
  background: #0ADC7D;
  border: 0;
  outline: 0;
  box-shadow: none;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem 0;
  padding-left: 1.6rem;
}

.c7:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 2.4rem #F4F4F4 inset;
  -webkit-filter: none;
  filter: none;
}

.c7:-webkit-autofill::first-line {
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
}

.c7:focus {
  border: 0;
}

.c2 {
  font-size: 1.4rem;
  color: #193241;
  cursor: pointer;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: #193241;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
}

.c5 > svg {
  width: 2.2rem;
  height: 100%;
}

.c5 span {
  padding: 0;
}

.c9 {
  color: #EE1010;
  font-size: 1.2rem;
}

.c0 .c3 {
  border-color: #EE1010;
}

.c0 .c1 {
  color: #EE1010;
}

<div
  class="c0"
>
  <label
    class="c1 c2"
  >
    TextField
  </label>
  <div
    class="c3 c4"
  >
    <div
      class="c5"
    >
      <svg
        aria-hidden="true"
        class="c6"
        data-testid="icon"
        fill="none"
        focusable="false"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 4h16c1.1 0 2 .9 2 2v12c0 1.1-.9 2-2 2H4c-1.1 0-2-.9-2-2V6c0-1.1.9-2 2-2z"
        />
        <polyline
          points="22,6 12,13 2,6"
        />
      </svg>
    </div>
    <select
      class="c7"
    >
      <option
        class="c8"
        value="1"
      >
        1
      </option>
    </select>
  </div>
  <p
    class="c9"
  >
    Error message
  </p>
</div>
`;
